optuna_dashboard/ts/components/GraphRank.tsx

```diff
@@ -0,0 +1,342 @@
+  }
+
+  const rankPlotInfo = getRankPlotInfo(study, objectiveId, xParam, yParam)
+
+  useEffect(() => {
+    if (study != null) {
+      plotRank(rankPlotInfo, theme.palette.mode)
+    }
+  }, [study, theme.palette.mode])
+
+  const space: SearchSpaceItem[] = study ? study.union_search_space : []
+
+  return (
+    <Grid container direction="row">
+      <Grid
+        item
+        xs={3}
+        container
+        direction="column"
+        sx={{ paddingRight: theme.spacing(2) }}
+      >
+        <Typography
+          variant="h6"
+          sx={{ margin: "1em 0", fontWeight: theme.typography.fontWeightBold }}
+        >
+          Rank
+        </Typography>
+        {study !== null && study.directions.length !== 1 ? (
+          <FormControl component="fieldset">
+            <FormLabel component="legend">Objective:</FormLabel>
+            <Select value={objectiveId} onChange={handleObjectiveChange}>
+              {study.directions.map((d, i) => (
+                <MenuItem value={i} key={i}>
+                  {objectiveNames.length === study?.directions.length
+                    ? objectiveNames[i]
+                    : `${i}`}
+                </MenuItem>
+              ))}
+            </Select>
+          </FormControl>
+        ) : null}
+        {study !== null && space.length > 0 ? (
+          <Grid container direction="column" gap={1}>
+            <FormControl component="fieldset" fullWidth>
+              <FormLabel component="legend">x:</FormLabel>
+              <Select value={xParam?.name || ""} onChange={handleXParamChange}>
+                {space.map((d) => (
+                  <MenuItem value={d.name} key={d.name}>
+                    {d.name}
+                  </MenuItem>
+                ))}
+              </Select>
+            </FormControl>
+            <FormControl component="fieldset" fullWidth>
+              <FormLabel component="legend">y:</FormLabel>
+              <Select value={yParam?.name || ""} onChange={handleYParamChange}>
+                {space.map((d) => (
+                  <MenuItem value={d.name} key={d.name}>
+                    {d.name}
+                  </MenuItem>
+                ))}
+              </Select>
+            </FormControl>
+          </Grid>
+        ) : null}
+      </Grid>
+      <Grid item xs={9}>
+        <Box id={plotDomId} sx={{ height: "450px" }} />
+      </Grid>
+    </Grid>
+  )
+}
+
+const getRankPlotInfo = (
+  study: StudyDetail | null,
+  objectiveId: number,
+  xParam: SearchSpaceItem | null,
+  yParam: SearchSpaceItem | null
+): RankPlotInfo | null => {
+  if (study === null) {
+    return null
+  }
+
+  const trials = study.trials
+  const filtered_trials = trials.filter(filterFunc)
+  if (filtered_trials.length < 2 || xParam == null || yParam == null) {
+    return null
+  }
+
+  const xAxis = getAxisInfo(filtered_trials, xParam)
+  const yAxis = getAxisInfo(filtered_trials, yParam)
+
+  const xValues: number[] = []
+  const yValues: number[] = []
```

Comment:
Do xValues and yValues support categorical values? In my local, the plot with categorical values seems not work well.

Reply:
I fixed the type of the variables. Thanks for the comment.

I will check the behavior for the categorical variables.

Reply:
Comparing the other plot, the rank plot has not supported categorical values yet.
<img width="1412" alt="スクリーンショット 2023-10-30 23 54 16" src="https://github.com/optuna/optuna-dashboard/assets/1277089/25e82707-e208-41bd-aa94-60c57f632bbd">
<img width="1411" alt="スクリーンショット 2023-10-30 23 54 23" src="https://github.com/optuna/optuna-dashboard/assets/1277089/82a1b992-d3af-4a0f-9c88-f3e188c8bbf2">

Reply:
I fixed the code to support the categorical variables. Basically, the order of the ticks is in dictionary order.

<img width="1818" alt="Screenshot 2023-10-31 at 14 46 04" src="https://github.com/optuna/optuna-dashboard/assets/38826298/05731707-9601-469f-88ce-f5de0be2a5fd">

If both the x-axis and y-axis are categorical variables, and there is no trial that yields the minimum values in dictionary order, the order of the ticks will not be in dictionary order. This is because we don't have a way to sort the ticks themselves. I believe we should accept this imperfection.

<img width="1839" alt="Screenshot 2023-10-31 at 14 49 58" src="https://github.com/optuna/optuna-dashboard/assets/38826298/9e79c3ef-6393-486d-a0ac-e4d8826df2df">